text/0000-action-creators.md

```diff
@@ -0,0 +1,107 @@
+- Start Date: 2018-02-09
+- RFC PR: (leave this empty)
+- React Navigation Issue: (leave this empty)
+
+# Modular Actions Creators
+
+Different routers should expose different actions. For example, it may make sense to 'pop' within a StackRouter, but an app with a drawer router should have `openDrawer` and `closeDrawer` actions available.
+
+AddNavigationHelpers forces `createNavigator` to support every action that every router supports, which prevents custom routers from exposing new action types. This is a proposal to replace `addNavigationHelpers` with a more modular set of navigation actions, defined by each router. This also allows us to deprecate the global concept of "NavigationActions" entirely.
+
+## Basic Example
+
+```
+// Only available in a Stack:
+this.props.navigation.popToTop();
+
+// Only available in a Drawer:
+this.props.navigation.openDrawer();
+```
+
+## Detailed design
+
+We will add a new function on the router, called "getActionsForRoute", which will return action creators. For example, in the StackRouter:
+
+```
+const stackRouter = {
+    getActionsForRoute: route => ({
+        popToTop: () => ({ type: 'PopToTop' }),
+        setParams: (params) => ({ type: 'SetParams', key: route.key, params }),
+        ...
+    }),
+    ...
+}
+```
+
+Inside of createNavigator, when the child navigation prop gets prepared, it will use the router to get the action creators. It will make the actions available within `childNavigation.actions`, and will also merge in parent actions provided via `this.props.navigation.actions`. That will enable the following:
+
+```
+// this.props.navigation.actions will only have `pop` within stacks:
+const action = this.props.navigation.actions.pop();
+this.props.navigation.dispatch(action);
+```
+
+As a convenience, the actions are also exposed on the navigation prop. These will call `dispatch` under the hood automatically:
+
+```
+// Only available in a Stack:
+this.props.navigation.popToTop();
+// Only available in a Drawer:
+this.props.navigation.openDrawer();
+// Only available in Tabs or Drawer:
+this.props.navigation.jumpTo();
+```
+
+This allows us to have the convenience of `addNavigationHelpers`, but each router can provide contextual helpers.
+
+
+## Action Creator Definitions
+
+Now is a good time to revisit the helpers that we have available, and what arguments they should.
+
+### Stack Router
+    - Navigate
+    - GoBack
+    - Push
+    - Pop
+    - PopToTop
+    - Reset
+
+### Tab Router
+    - Navigate
+    - GoBack
+    - JumpTo
+
+### Drawer Router
+    - Navigate
+    - GoBack
+    - JumpTo
+    - OpenDrawer
+    - CloweDrawer
+    - ToggleDrawer
+
+
+### Drawbacks
+
+The action creators will override eachother in the same namespace. Which means that if you nest two stack navigators, you will only be able to access the action creators of the inner navigator when calling from inside it.
```

Comment:
> Can you use the parent’s action creator to navigate on the grandparent?

Yes, exactly! As long as the action creator is compatible. So, we will need to make sure, for example, that StackRouter and TabRouter both expose an identical navigation action and back action creator, so that they can nest without strange behavior. For people with custom routers, they will need to make sure their actions names don't conflict with the built-in routers.

So there is a bit of complexity here, but I think it will result in predictable and debuggable behavior.